README.md
<h1 align="center">
  <code>monetization</code>
</h1>
<h3 align="center">
  A wrapper around the <a href="https://webmonetization.org/">Web Monetization API</a>.
</h3>

<p align="center">
    <a href="https://pub.dev/packages/monetization">
        <img src="https://img.shields.io/pub/v/monetization?style=for-the-badge" title="Pub Version" />
    </a>
    &nbsp;
    <a href="./LICENSE">
        <img src="https://img.shields.io/badge/license-MIT-green.svg?style=for-the-badge" title="License" />
    </a>
</p>

Offer extra content and features for users who stream micro-payments — including premium features, additional content, or digital goods.

<h3 align="center">
  <a href="https://pub.dev/documentation/monetization/latest/monetization/Monetization-class.html">API Reference</a>
</h3>

## Usage

A simple usage example that initializes the monetization to a specific payment pointer:

```dart
import 'package:monetization/monetization.dart';

main() {
  var monetization = Monetization('\$pay.tomasarias.me');
}
```

You can subscribe to Web Monetization events using a `Stream`:

```dart
monetization.onPending.listen((event) {
  // Prepare to serve the monetized content
});

monetization.onStart.listen((event) {
  // Show monetized content
});

monetization.onProgress.listen((event) {
  // Do something on each micro-payment
});

monetization.onStop.listen((event) {
  // Hide monetized content
});
```

You can also check if a user is paying without subscribing to the streams:

```dart
Future<bool> isPaying() async {
  // Prefer custom logic over this
  await Future.delayed(const Duration(seconds: 3));
  return monetization.isPaying;
}
```

### Get information about the monetization

```dart
monetization.isMonetized; // Returns if the user supports monetization
monetization.isPaying;    // Returns if the user is streaming payments
monetization.pointer;     // Returns the current payment pointer
```

### Get the revenue from the current session

```dart
monetization.getTotal(formatted: false); // 884389
monetization.getTotal(); // 0.000884389
monetization.assetCode;  // 'XRP'
monetization.assetScale; // 9
```

### Enable/disable the monetization dynamically 

```dart
monetization.enabled;   // true
monetization.disable(); // Stops the monetization
monetization.enable();  // Start the monetization again with the same pointer
```

### Probabilistic revenue sharing

Sometimes you want to share revenue across different people, to do this pass a `Map` with the payment pointers and
weights to the `Monetization.probabilistic` constructor.

This will choose one of the pointers for the entire session.

```dart
final pointers = {
  'pay.tomasarias.me/usd': 0.5,
  'pay.tomasarias.me/xrp': 0.2,
  'pay.tomasarias.me/ars': 0.3
};

var monetization = Monetization.probabilistic(pointers);
```

This will result on a 50% chance of choosing the first pointer, 20% chance of choosing the second and 30% chance of
choosing the third one.

For more information on probabilistic revenue sharing, read [this article](https://coil.com/p/sharafian/Probabilistic-Revenue-Sharing/8aQDSPsw)
by Ben Sharafian.

### Verifying payments (using [Vanilla](https://vanilla.so))

Monetization events can be manipulated, so you can't know for sure if a user is really paying. You can add an extra layer of security to your monetized content using Vanilla. To learn how does Vanilla works read
[this article](https://dev.to/cinnamonvideo/vanilla-by-cinnamon-497).

You will need your API credentials, you can get them [here](https://admin.vanilla.so).

```dart
// Vanilla API Credentials
final clientId = 'Your Client ID';
final clientSecret = 'Your Client Secret';

var monetization = Monetization.vanilla(clientId, clientSecret);
```

Now you can check if the payment stream is valid in different ways:

```dart
// With Vanilla, this will return true if the user is paying and Vanilla generates a proof of payment.
monetization.isPaying;
// With Vanilla, this will return the current payment rate per second, if the monetization is stopped this will be 0.
monetization.getVanillaRate()
// With Vanilla, this will return the total amount received from the current requestId.
monetization.getVanillaTotal();
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker](https://github.com/KNawm/monetization/issues).


CHANGELOG.md
## 1.2.0

- Added support for [Vanilla](https://vanilla.so)
- Added `enabled` getter that returns if the meta tag is present
- `getTotal()` now returns the formatted amount by default
- Example now uses Vanilla
- Improved docs

## 1.0.1

- Format and pubspec changes.

## 1.0.0

- Initial release.


test/monetization_test.dart
import 'package:monetization/monetization.dart';
import 'package:test/test.dart';

void main() {
  group('Probabilistic', () {
    Monetization wm;
    int a, b, c;

    final pointers = { 'pay.tomasarias.me/usd': 0.5,
      'pay.tomasarias.me/xrp': 0.2,
      'pay.tomasarias.me/ars': 0.3
    };

    setUp(() {
      for (int i = 0; i < 5000; i++) {
        wm = Monetization.probabilistic(pointers);

        if (wm.pointer == pointers.keys.elementAt(0)) {
          a++;
        } else if (wm.pointer == pointers.keys.elementAt(1)) {
          b++;
        } else if (wm.pointer == pointers.keys.elementAt(2)) {
          c++;
        }
      }
    });

    test('Simulation', () {
      expect(a / 5000, closeTo(pointers[0], 0.1));
      expect(b / 5000, closeTo(pointers[1], 0.1));
      expect(c / 5000, closeTo(pointers[2], 0.1));
    });
  });
}


lib/src/widget.dart
/*import 'package:flutter/material.dart';

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'this is the web',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => addMetaTag(),
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

class Monetized extends StatefulWidget {
  Monetized({Key key, /*this.title*/}) : super(key: key);

  @override
  _MonetizedState createState() => _MonetizedState();
}

class _MonetizedState extends State<Monetized> {
  @override
  Widget build(BuildContext context) {
    return Container();
  }
*/

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:monetization/monetization.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter + Web Monetization',
      home: MyHomePage(title: 'Web Monetization Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Monetization monetization;
  double total;

  // Vanilla API Credentials
  final clientId = const String.fromEnvironment('CLIENT_ID');
  final clientSecret = const String.fromEnvironment('CLIENT_SECRET');

  @override
  void initState() {
    super.initState();
    setState(() {
      monetization = Monetization.vanilla(clientId, clientSecret, debug: true);
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        centerTitle: true,
        backgroundColor: Color(0xFF6ADAAB),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Padding(
              padding: const EdgeInsets.all(16.0),
              child: Image(
                image: AssetImage('assets/wm.png'),
                width: 275,
                height: 275,
              ),
            ),
            StreamBuilder(
                stream: monetization.onProgress,
                builder: (context, snapshot) {
                  if (!snapshot.hasData) {
                    return Text(
                      'Monetization not started yet',
                      style: Theme.of(context).textTheme.headline4,
                    );
                  }

                  return Text(
                    '${monetization.getTotal()} ${monetization.assetCode}',
                    style: Theme.of(context).textTheme.headline4,
                  );
                }),
            Padding(
              padding: const EdgeInsets.all(24.0),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  RaisedButton(
                    onPressed: () => monetization.enable(),
                    child: Text(
                      'Start',
                      style: Theme.of(context).textTheme.headline6,
                    ),
                  ),
                  SizedBox(
                    width: 24,
                  ),
                  RaisedButton(
                    onPressed: () => monetization.disable(),
                    child: Text(
                      'Stop',
                      style: Theme.of(context).textTheme.headline6,
                    ),
                  ),
                ],
              ),
            ),
            Text(
              'Tip: Open the DevTools to see the monetization events.',
              style: Theme.of(context).textTheme.overline,
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/monetization_base.dart
import 'dart:convert' show base64Encode, jsonEncode, jsonDecode, utf8;
import 'dart:html'
    show CustomEvent, EventStreamProvider, EventTarget, MetaElement, document;
import 'dart:math' as math show pow, Random;
import 'package:http/http.dart' as http;
import 'interop.dart' as js;

class Monetization {
  // Monetization DOM object
  EventTarget _monetization;
  // Current payment pointer stored in memory
  String _paymentPointer;
  String _assetCode;
  int _assetScale;
  String _state;
  double _total;
  // Monetization events
  Stream<CustomEvent> _onPending;
  Stream<CustomEvent> _onStart;
  Stream<CustomEvent> _onStop;
  Stream<CustomEvent> _onProgress;
  // Vanilla-related
  String _vanillaAuth;
  double _vanillaRate;
  double _vanillaTotal;
  http.Client _vanillaClient;
  bool get _vanilla => _vanillaAuth != null;

  /// Returns whether the meta tag is set to receive payments.
  bool get enabled => _getMetaTag() != null;

  /// Returns whether the user supports Web Monetization.
  bool get isMonetized => state != 'undefined';

  /// Returns whether the user is streaming payments.
  bool get isPaying {
    if (_vanilla) {
      // Is paying if the rate from the last payment proof is greater than 0.
      return _vanillaRate > 0;
    }

    return state == 'started';
  }

  /// Returns the current payment pointer.
  ///
  /// To check if a meta tag with this pointer is set use [enabled].
  String get pointer => _paymentPointer;

  /// Returns the code identifying the asset's unit.
  ///
  /// For example: 'USD' or 'XRP'.
  String get assetCode => _assetCode;

  /// Returns the number of places past the decimal for the amount.
  ///
  /// For example, if you have USD with an [assetScale] of 2, then the minimum
  /// divisible unit is cents.
  int get assetScale => _assetScale;

  /// Returns the monetization state provided by the browser.
  ///
  /// **`undefined`**:
  /// Monetization is not supported for this user.
  ///
  /// **`pending`**:
  /// Streaming has been initiated, yet first non zero packet is
  /// "pending". It will normally transition from this `state` to `started`,
  /// yet not always.
  ///
  /// **`started`**:
  /// Streaming has received a non zero packet and is still active.
  ///
  /// **`stopped`**:
  /// Streaming is inactive. This could mean a variety of things:
  /// - May not have started yet
  /// - May be paused (potentially will be resumed)
  /// - Has finished completely (and awaits another request)
  /// - The payment request was denied by user intervention
  String get state => _state ?? 'undefined';

  /// Stream that tracks 'monetizationpending' events.
  ///
  /// This event fires when Web Monetization is enabled.
  Stream<Map> onPending;

  /// Stream that tracks 'monetizationstart' events.
  ///
  /// This event fires when Web Monetization has started actively paying.
  Stream<Map> onStart;

  /// Stream that tracks 'monetizationstop' events.
  ///
  /// This event fires when Web Monetization has stopped.
  Stream<Map> onStop;

  /// Stream that tracks 'monetizationprogress' events.
  ///
  /// This event fires when Web Monetization has streamed a payment.
  Stream<Map> onProgress;

  /// Returns whether to log events to the console.
  bool debug;

  /// Initialize Web Monetization supplying a [paymentPointer].
  factory Monetization(String paymentPointer, {debug = false}) {
    return Monetization._(paymentPointer, debug);
  }

  Monetization._(String paymentPointer, this.debug, [String auth]) {
    if (js.supportsMonetization) {
      _vanillaAuth = auth;
      _vanillaRate = 0;
      _setPaymentPointer(paymentPointer);
      _monetization = js.monetization;
      _state = js.state;
      _addMetaTag();
      _addEventHandlers();
    }
  }

  /// Initialize Web Monetization supplying a Map of payment pointers with their
  /// respective weights.
  ///
  /// For example:
  /// ```
  /// var pointers = { 'pay.tomasarias.me/usd': 0.5,
  ///                  'pay.tomasarias.me/xrp': 0.2,
  ///                  'pay.tomasarias.me/ars': 0.3 };
  /// ```
  ///
  /// For more information,
  /// see <https://coil.com/p/sharafian/Probabilistic-Revenue-Sharing/8aQDSPsw>
  factory Monetization.probabilistic(Map<String, double> paymentPointers,
      {debug = false}) {
    final sum = paymentPointers.values.reduce((sum, weight) => sum + weight);
    var choice = math.Random().nextDouble() * sum;
    String paymentPointer;

    for (final pointer in paymentPointers.keys) {
      final weight = paymentPointers[pointer];
      if ((choice -= weight) <= 0) {
        paymentPointer = pointer;
      }
    }

    return Monetization._(paymentPointer, debug);
  }

  /// Initialize Web Monetization using Vanilla.
  ///
  /// For more information,
  /// see <https://vanilla.so>
  factory Monetization.vanilla(String clientId, String clientSecret,
      {debug = false}) {
    final auth = base64Encode(utf8.encode('$clientId:$clientSecret'));
    return Monetization._('\$wm.vanilla.so/pay/$clientId', debug, auth);
  }

  /// Enable Web Monetization.
  void enable() {
    _addMetaTag();
  }

  /// Disable Web Monetization.
  void disable() {
    _removeMetaTag();
  }

  /// Returns the amount received by the current [pointer] on this session.
  ///
  /// Keep in mind that this number resets every time that `onStart` fires.
  ///
  /// Pass `formatted: false` if you want the "raw" total (without accounting for the scale of the asset).
  double getTotal({bool formatted = true}) {
    if (formatted) {
      return double.parse(
          (_total * math.pow(10, -_assetScale)).toStringAsFixed(_assetScale));
    }

    return _total;
  }

  /// Returns the current payment rate per second from Vanilla's proof of
  /// payment on "raw" format (disregarding the asset scale).
  double getVanillaRate() => _vanillaRate;

  /// Returns the total amount received from Vanilla's proof of payment in "raw"
  /// format (disregarding the asset scale).
  ///
  /// This total SHOULD be the same that you get from `getTotal(formatted: false)`
  /// but sometimes there will be a small discrepancy.
  ///
  /// Keep in mind this total is only from the last requestId and will reset
  /// if the requestId changes.
  double getVanillaTotal() => _vanillaTotal;

  void _setPaymentPointer(String paymentPointer) {
    if (paymentPointer.startsWith('\$')) {
      _paymentPointer = paymentPointer;
    } else {
      throw ArgumentError.value(paymentPointer, 'paymentPointer',
          'The payment pointer must start with "\$"');
    }
  }

  MetaElement _getMetaTag() {
    return document.head.querySelector('meta[name="monetization"]');
  }

  void _addMetaTag() {
    if (_getMetaTag() == null) {
      final metaTag = MetaElement();
      metaTag.name = 'monetization';
      metaTag.content = _paymentPointer;
      document.head.append(metaTag);
    } else {
      _updateMetaTag(_paymentPointer);
    }
  }

  void _updateMetaTag(String paymentPointer) {
    final metaTag = _getMetaTag();

    if (metaTag != null) {
      metaTag.setAttribute('content', paymentPointer);
    }
  }

  void _removeMetaTag() {
    final metaTag = _getMetaTag();

    if (metaTag != null) {
      metaTag.remove();
    }
  }

  Future<String> _proof(String requestId) async {
    if (_vanilla) {
      Map<String, dynamic> proof;
      final query = '''
      {
        proof(requestId: "$requestId") {
          rate
          total
        }
      }
      ''';

      final headers = {
        'Content-type': 'application/json',
        'Authorization': 'Basic $_vanillaAuth'
      };
      try {
        final response = await _vanillaClient.post(
            'https://wm.vanilla.so/graphql',
            body: jsonEncode({'query': query}),
            headers: headers);

        proof = jsonDecode(response.body);
        _vanillaRate = proof['data']['proof']['rate'];
        _vanillaTotal = proof['data']['proof']['total'];

        return proof.toString();
      } catch (e, s) {
        if (debug) {
          js.log('Error getting payment proof:\n$e\n$s');
        }
      }
    }

    return null;
  }

  Future<void> _monetizationPending(CustomEvent event) async {
    _state = js.state;
    _debug(event);
  }

  Future<void> _monetizationStart(CustomEvent event) async {
    _total = 0;
    _vanillaClient = http.Client();
    _state = js.state;
    _debug(event);
  }

  Future<void> _monetizationStop(CustomEvent event) async {
    _vanillaClient.close();
    _vanillaRate = 0;
    _state = js.state;
    _debug(event);
  }

  Future<void> _monetizationProgress(CustomEvent event) async {
    final requestId = event.detail['requestId'];

    if (_total == 0) {
      _assetCode = event.detail['assetCode'];
      _assetScale = event.detail['assetScale'];
    }

    _total += double.parse(event.detail['amount']);
    _state = js.state;
    _debug(event, await _proof(requestId));
  }

  void _addEventHandlers() {
    _onPending = const EventStreamProvider<CustomEvent>('monetizationpending')
        .forTarget(_monetization);
    _onPending.listen((event) {
      _monetizationPending(event);
    });
    onPending = _onPending.map((CustomEvent event) => _mapify(event));

    _onStart = const EventStreamProvider<CustomEvent>('monetizationstart')
        .forTarget(_monetization);
    _onStart.listen((event) {
      _monetizationStart(event);
    });
    onStart = _onStart.map((CustomEvent event) => _mapify(event));

    _onStop = const EventStreamProvider<CustomEvent>('monetizationstop')
        .forTarget(_monetization);
    _onStop.listen((event) {
      _monetizationStop(event);
    });
    onStop = _onStop.map((CustomEvent event) => _mapify(event));

    _onProgress = const EventStreamProvider<CustomEvent>('monetizationprogress')
        .forTarget(_monetization);
    _onProgress.listen((event) {
      _monetizationProgress(event);
    });
    onProgress = _onProgress.map((CustomEvent event) => _mapify(event));
  }

  Map<String, dynamic> _mapify(CustomEvent event) {
    final map = <String, dynamic>{
      'type': js.get(event, 'type'),
      'timeStamp': js.get(event, 'timeStamp'),
    };
    map.addAll((event.detail as Map).cast<String, dynamic>());

    return map;
  }

  void _debug(CustomEvent event, [String proof]) {
    if (debug) {
      final map = _mapify(event);

      if (map['type'] == 'monetizationprogress') {
        js.log('${map.toString()}');
      } else if (map['type'] == 'monetizationpending') {
        js.log('%c${map.toString()}', 'color: yellow');
      } else if (map['type'] == 'monetizationstart') {
        js.log('%c${map.toString()}', 'color: lime');
      } else {
        js.log('%c${map.toString()}', 'color: red');
      }

      if (proof != null) {
        js.log(proof);
      }
    }
  }
}
